kvserver: wake leader on Raft message from other replica

erikgrinaker opened the pull request:
**kvserver: wake leader on Raft message from other replica**

Previously, in the case of a quiesced range, a follower who received a pre-vote request could unquiesce itself only to reject the pre-vote. Under a partial network partition, where the leader itself did not receive a corresponding pre-vote request, we would now have two unquiesced followers with a quiesced leader. Eventually, both of the followers' election timeouts would lapse and they would hold an election, upsetting the established leader. Furthermore, the leadership would often be immediately transferred back to the leaseholder, i.e. the old leader, and the cycle would repeat.

The same holds for any other message that a follower might send to a different follower, without also sending a message to the leader.

This patch makes a quiesced follower unquiesce both itself and the leader when it receives a message from anyone but the leader, allowing the leader to maintain leadership.

Note that both before and after this patch, such partial partitions often result in persistent mass unquiescence due to the continuous prevotes.

Touches #92088.

Epic: None.
Release note: None.

**kvserver: don't unquiesce during snapshot application**

Previously, snapshot application would explicitly unquiesce the newly initialized replica. `handleRaftReady` would also unquiesce the replica once the snapshot is applied, but the former did so without also waking the leader. This can lead to spurious elections if multiple replicas are left unquiesced without waking the leader, e.g. during a partial network partition.

This patch therefore removes the explicit unquiesce during snapshot application, and instead unquiesces normally (along with the leader) during `handleRaftReady`. This is also symmetric with other replica initialization, which does not explicitly unquiesce the replica either.

Touches #92088.

Epic: none
Release note: None

erikgrinaker committed:
kvserver: add Raft PreVote and CheckQuorum tests

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104057)
<!-- Reviewable:end -->

erikgrinaker committed:
kvserver: wake leader on Raft message from other replica

Previously, in the case of a quiesced range, a follower who received a
pre-vote request could unquiesce itself only to reject the pre-vote.
Under a partial network partition, where the leader itself did not
receive a corresponding pre-vote request, we would now have two
unquiesced followers with a quiesced leader. Eventually, both of the
followers' election timeouts would lapse and they would hold an
election, upsetting the established leader. Furthermore, the leadership
would often be immediately transferred back to the leaseholder, i.e.
the old leader, and the cycle would repeat.

The same holds for any other message that a follower might send to a
different follower, without also sending a message to the leader.

This patch makes a quiesced follower unquiesce both itself and the
leader when it receives a message from anyone but the leader, allowing
the leader to maintain leadership.

Note that both before and after this patch, such partial partitions
often result in persistent mass unquiescence due to the continuous
prevotes.

Epic: none
Release note: None

erikgrinaker commented:
On second thought, I think we may want to always wake the leader whenever a follower unquiesces, since it will otherwise always result in the replica campaigning.

erikgrinaker commented:
Changed this to wake the leader on any message from a non-leader replica, instead of only on prevotes.

erikgrinaker commented:
I think we may already have the mass unquiescence problem. If the follower who is not partitioned wins the election (it is likely to have the longest log), then it will immediate transfer leadership back to the leaseholder (the original leader). This cycle will then likely continue indefinitely. Worth confirming, but if so, I'm less anxious about enabling this by default.

erikgrinaker commented:
>I think we may already have the mass unquiescence problem.

Confirmed. Prior to this PR, with a partial partition, there's generally two cases: either the partitioned follower is able to continue heartbeating liveness, in which case the leader won't quiesce because the follower isn't caught up, or the partitioned follower will wake the unpartitioned follower via prevote, eventually resulting in an election, where the leadership gets transferred back to the leaseholder (previous leader). This goes on and on.

As such, this appears to be a strict improvement, so I've removed the `CheckQuorum` gate here.

erikgrinaker committed:
kvserver: don't unquiesce during snapshot application

Previously, snapshot application would explicitly unquiesce the newly
initialized replica. `handleRaftReady` would also unquiesce the replica
once the snapshot is applied, but the former did so without also waking
the leader. This can lead to spurious elections if multiple replicas are
left unquiesced without waking the leader, e.g. during a partial network
partition.

This patch therefore removes the explicit unquiesce during snapshot
application, and instead unquiesces normally (along with the leader)
during `handleRaftReady`. This is also symmetric with other replica
initialization, which does not explicitly unquiesce the replica either.

Epic: none
Release note: None

pavelkalinnikov commented on pkg/kv/kvserver/client_raft_test.go:
Why 3 seconds? If you put 10 seconds here, would it still stay in the same state? Just checking that there isn't a race here with some fixed duration.

erikgrinaker replied:
Correct. We just have to wait "long enough" that the replica won't transition into Candidate or Follower, either because of any received messages or because of an election timeout. Updated the comment to clarify this.

erikgrinaker commented:
TFTR!

bors r+

erikgrinaker commented:
bors retry
